### DISCUSSION TITLE
"Deleting Objects Referenced from Collections and Scalar Relationships" - Async Edition

### DISCUSSION THREAD

**Discussion by bfontaine** [sqlalchemy/sqlalchemy#9756] [Usage Questions] [answered]
Hello,
In the documentation, the section "[Notes on Delete - Deleting Objects Referenced from Collections and Scalar Relationships](https://docs.sqlalchemy.org/en/20/orm/cascades.html#notes-on-delete-deleting-objects-referenced-from-collections-and-scalar-relationships)" advices to use code like `del user.addresses[1]` to avoid issues with `Session.delete()`, but this does not work with async code as `async del ...` makes no sense.
Do you have some guidance on how to do that?

In my specific case I use SQLAlchemy 2.0 with async and SQLite (for tests; in production I use Postgres) have a situation like this:

```python
# models:
class Parent(Base):
    id: Mapped[int] = mapped_column(Integer, primary_key=True)
    children: Mapped[List["Child"]] = relationship("Child", back_populates="parent")

class Child(Base):
    id: Mapped[int] = mapped_column(Integer, primary_key=True)
    name: Mapped[str] = mapped_column(String, nullable=False, unique=True)

    parent_id: Mapped[int] = mapped_column(Integer, ForeignKey("children.id", ondelete="CASCADE"),  nullable=False)
    parent: Mapped[Parent] = relationship("Parent", back_populates="children", cascade="all", lazy="selectin")

# code:
parent1 = await session.execute(select(Parent).where(Parent.id == 1)).unique().scalar_one()
parent2 = await session.execute(select(Parent).where(Parent.id == 2)).unique().scalar_one()

bob = await session.execute(select(Child).where(Child.name == "Bob")).unique().scalar_one()
print(bob.parent.id)  # 1

await session.delete(bob)
await session.flush()
new_bob = Child(name="Bob", parent=parent2)
session.add(new_bob)
await session.commit()
```
However I get this error:
> Instance '<Parent at 0x7fb772775570>' has been deleted.  Use the make_transient() function to send this object back to the transient state.

I don’t understand why the error says the parent was deleted because I only wants to delete the child.

Update: if I use `await session.execute(delete(Child).where(Child.id == bob.id))` instead of `await session.delete(bob)`, it doesn’t delete the parent.

**Comment by zzzeek** [comment, marked as answer]
> Hello, In the documentation, the section "[Notes on Delete - Deleting Objects Referenced from Collections and Scalar Relationships](https://docs.sqlalchemy.org/en/20/orm/cascades.html#notes-on-delete-deleting-objects-referenced-from-collections-and-scalar-relationships)" advices to use code like `del user.addresses[1]` to avoid issues with `Session.delete()`, but this does not work with async code as `async del ...` makes no sense.

That's the Python `del` keyword operating on a collection (a Python list),  it's not an `async` operation for a loaded collection.   The usual rules for accessing collections under asyncio apply, where you would want to make sure that collection was loaded via eager loading, and in the next release of SQLAlchemy we will also have a helper object so that you can also lazy load this (in this case, using `del` and assuming the collection were not loaded, it would look like `del (await user.awaitable_attrs.addresses)[1]`).


> Do you have some guidance on how to do that?
> 
> In my specific case I use SQLAlchemy 2.0 with async and SQLite (for tests; in production I use Postgres) have a situation like this:

There are some mistakes in your code like "Child.parent_id" referring to "children", if I fix the mistakes no such error is generated, even though it's a little unusual to have "cascade='all'" on Child.parent ("all" includes "delete", so that it means if you delete a Child, its Parent is also deleted, which seems like not what you'd want here, perhaps in your actual program that's what the problem is).

as written the code completes because "bob" and its parent(id=1) is deleted, parent(id=2) is fine and you get the new INSERT.

> Update: if I use await session.execute(delete(Child).where(Child.id == bob.id)) instead of await session.delete(bob), it doesn’t delete the parent.

OK, so I guess the "delete of the parent" is the problem, just take that "cascade='all'" away, which doesn't really belong on a many-to-one, see full case below

```py
from __future__ import annotations

from typing import List

from sqlalchemy.ext.asyncio import create_async_engine
from sqlalchemy.ext.asyncio import AsyncSession
from sqlalchemy import create_engine, select
from sqlalchemy import ForeignKey, Integer, String
from sqlalchemy.orm import DeclarativeBase
from sqlalchemy.orm import Mapped
from sqlalchemy.orm import mapped_column
from sqlalchemy.orm import relationship
from sqlalchemy.orm import Session


class Base(DeclarativeBase):
    pass


class Parent(Base):
    __tablename__ = 'parent'
    id: Mapped[int] = mapped_column(Integer, primary_key=True)
    children: Mapped[List["Child"]] = relationship(
        "Child", back_populates="parent"
    )


class Child(Base):
    __tablename__ = 'child'
    id: Mapped[int] = mapped_column(Integer, primary_key=True)
    name: Mapped[str] = mapped_column(String, nullable=False, unique=True)

    parent_id: Mapped[int] = mapped_column(
        Integer, ForeignKey("parent.id", ondelete="CASCADE"), nullable=False
    )
    parent: Mapped[Parent] = relationship(
        "Parent", back_populates="children", lazy="selectin"
    )


async def main():
    e = create_async_engine("sqlite+aiosqlite://", echo=True)
    async with e.begin() as conn:
        await conn.run_sync(Base.metadata.create_all)

    session = AsyncSession(e)

    session.add_all(
        [
            Parent(id=1, children=[Child(name="Bob")]),
            Parent(
                id=2,
            ),
        ]
    )
    await session.commit()

    # code:
    parent1 = (
        (await session.execute(select(Parent).where(Parent.id == 1)))
        .unique()
        .scalar_one()
    )
    parent2 = (
        (await session.execute(select(Parent).where(Parent.id == 2)))
        .unique()
        .scalar_one()
    )

    bob = (
        (await session.execute(select(Child).where(Child.name == "Bob")))
        .unique()
        .scalar_one()
    )
    print(bob.parent.id)  # 1

    await session.delete(bob)
    await session.flush()
    new_bob = Child(name="Bob", parent=parent2)
    session.add(new_bob)
    await session.commit()

    # not deleted
    assert parent1 in session

import asyncio

asyncio.run(main())

```





```py

**Reply by bfontaine** [reply]
Aaah thank you very much for the detailed answer! I mixed things up and thought the `cascade=` was meant to be used on the child to specify what happens to it when its parent is updated/deleted, but in reality it’s the opposite.

> That's the Python `del` keyword operating on a collection (a Python list), it's not an `async` operation for a loaded collection

Yes, my confusion came from the fact that when I did `del parent.children[0]` I had a `sqlalchemy.exc.MissingGreenlet` error, and assumed it was because the `del ...` somehow triggered other (async) operations behind the scenes. In reality, the error has nothing to do with `del` but with accessing `parent.children`.
